docs/DeveloperGuide: improve explanation for `find` feature. (#172)
CS2113-AY1819S2-T09-1/main: docs/DeveloperGuide.adoc

> The find feature aims to help users to be able to locate any module in our application easily. We support

Reviewer: I think it will be better to change this part too. 
```suggestion
The find feature aims to help users to be able to easily locate any module in our application. 
```

> The find feature aims to help users to be able to locate any module in our application easily. We support

Reviewer 2: I think it will be better to change this part too. 
```suggestion
The find feature aims to help users to be able to easily locate any modules in our application. 
```

> the finding of module's name, code and credits. This enable our users to be able to find for any modules with partial

Reviewer 2: ```suggestion
We support the finding of modules with their names, codes and credits. This enables our users to find any modules with partial
```

>  knowledge about the module they need.

Reviewer 2: Do you mean this?
```suggestion
information of the modules they have.
```

Step 1 is performed by the `ApplicationParser` class, and no special actions is needed for the `find` feature.

> Step 2 and 3 is performed by `BooleanExpression#parse`

Reviewer 3: Should be `BooleanExpressionParser#parse(...)`

The main implementation of this feature is split into two components. The `Tokenizer` and `BooleanExpressionParser`

> `FindCommandParser` calls `BooleanExpressionParser#parse` which handle the evaluation of the expression.

Reviewer 2: ```suggestion
`FindCommandParser` calls `BooleanExpressionParser#parse` which handles the evaluation of the expression.
```

1. `Tokenizer` helps to split the user provided argument into tokens which could be used by `BooleanExpressionParser`.

> 2. `BooleanExpressionParser` handle the high-level syntax of the find expressions and map them into

Reviewer 2: ```suggestion
2. `BooleanExpressionParser` handles the high-level syntax of the find expressions and maps them into
```

The main implementation of this feature is split into two components. The `Tokenizer` and `BooleanExpressionParser`

`FindCommandParser` handles the users provided input and ensure that the provided is not empty. Once it passes the

> checks, the argument are passed to `BooleanExpressionParser` which handle the evaluation of the expression.

Reviewer 3: The mixed use of `input` and `argument` makes it confusing for readers.
I think you could simplify the second sentence by saying `FindCommandParser` calls `BooleanExpressionParser#parse(...)` to evaluate the expression.

Each query will consume the previous token and returns the next available token. +
This is similar to how `java.util.Scanner` works.

===== Design Consideration

After many rounds of experiment with `StringTokenizer` provided by native Java and `ArgumentMultimap`. +

> We found two main issues which could not be satisfy by either `StringTokenizer` or `ArgumentMultimap`.

Reviewer 2: ```suggestion
We found two main issues which could not be satisfied by either `StringTokenizer` or `ArgumentMultimap`.
```

. `ArgumentMultimap` does not keep track of the order of each delimiter. They will only track if the delimiter provided
exists.

> . `StringTokenizer` default delimiter is whitespace, although we could change the delimiter and parse multiple

Reviewer 2: ```suggestion
. `StringTokenizer` has default delimiter as whitespace, although we could change the delimiter and parse multiple
```

delimiters. It does not suit our situation. +
e.g. `find name/AAAA code/BBBB`. +
This will return us 1 token. `name/AAAA code/BBBB`, which we could not use to check due to the missing `operator`. +
We need the tokenizer to return us 2 tokens `name/AAAA` and `code/BBBB` in order for us to know that the expression
was invalid due to the missing `operator`

As such, our custom `Tokenizer` aims to solve these 2 issues.

The table below shows the differences of our `Tokenizer` vs `StringTokenizer` and `ArgumentMultimap`
|====
| *`Tokenizer`* | *`StringTokenizer`* | *`ArgumentMultimap`*
| Respect the order of delimiters. | Respect the order of delimiters | Only keep track which delimiters is present.
| Support multiple delimiters. | Support only multiple delimiters. | Support multiple delimiters.
| More flexible in splitting of token | Only split token based on delimiters | Does not split into token.
|====

==== Operator

> This is represented by the class `pwe.planner.logic.parser.Operator` and define all valid operators to be used in

Reviewer 2: ```suggestion
This is represented by the class `pwe.planner.logic.parser.Operator` and defines all valid operators to be used in
```

Each query will consume the previous token and returns the next available token. +
This is similar to how `java.util.Scanner` works.

===== Design Consideration

> After many rounds of experiment with `StringTokenizer` provided by native Java and `ArgumentMultimap`. +

Reviewer 2: ```suggestion
After many rounds of experiment with `StringTokenizer` that is provided by native Java and `ArgumentMultimap`. +
```

We found three main issues which could not be satisfied by either `StringTokenizer` or `ArgumentMultimap`.

> . `ArgumentMultimap` does not keep track of the order of each delimiter. They will only track if the delimiter provided

Reviewer 2: ```suggestion
. `ArgumentMultimap` does not keep track of the order of each delimiter. They will only track if the delimiter is provided
```

exists.

> . `StringTokenizer` has default delimiter as whitespace, although we could change the delimiter and parse multiple

Reviewer 2: ```suggestion
. `StringTokenizer` has a default delimiter as a whitespace, although we could change the delimiter and parse multiple
```

delimiters. It does not suit our situation. +
e.g. `find name/AAAA code/BBBB`. +

> This will return us 1 token. `name/AAAA code/BBBB`, which we could not use to check due to the missing `operator`. +

Reviewer 2: ```suggestion
This will return us 1 token. `name/AAAA code/BBBB`, which we cannot use to check due to the missing `operator`. +
```
`cannot` since I believe it implies now too

We need the tokenizer to return us 2 tokens `name/AAAA` and `code/BBBB` in order for us to know that the expression
was invalid due to the missing `operator`.

> . `StringTokenizer` can take in `PREFIX` as delimiter, however, this will split the argument (i.e. `code/CS1231`) into

Reviewer 2: You may want to add a newline after line 431 just like you did at other lines.

two tokens, `code/` and `CS1231`. Additional parsing is required before we are able to pass it to

> `BooleanExpressionParser`. These parsing creates more overhead as we have to ensure there are proper checks is done.

Reviewer 2: ```suggestion
`BooleanExpressionParser`. This parsing creates more overhead as we have to ensure that proper checks are carried out.
```

As such, our custom `Tokenizer` aims to solve these three issues.

> The table below shows the differences of our `Tokenizer` vs `StringTokenizer` and `ArgumentMultimap`

Reviewer 2: ```suggestion
The table below shows the differences among our `Tokenizer`, `StringTokenizer` and `ArgumentMultimap`
```

|====
| *`Tokenizer`* | *`StringTokenizer`* | *`ArgumentMultimap`*

> | Respect the order of delimiters. | Respect the order of delimiters | Only keep track which delimiters is present.

Reviewer 2: ```suggestion
| Respect the order of delimiters. | Respect the order of delimiters | Only keep track which delimiters are present.
```

> | Support multiple delimiters. | Support only multiple delimiters. | Support multiple delimiters.

Reviewer 2: ```suggestion
| Support multiple delimiters. | Support multiple delimiters. | Support multiple delimiters.
```
Is there any reason why is there `only` in just that one?

> | More flexible in splitting of token | Only split token based on delimiters | Does not split into token.

Reviewer 2: ```suggestion
| Splitting tokens is more flexible | Only split token based on delimiters | Does not split into token.
```

|====

==== Operator

This is represented by the class `pwe.planner.logic.parser.Operator` and defines all valid operators to be used in
`BooleanExpressionParser`.

> The table below shows the valid operators our application currently support.

Reviewer 2: ```suggestion
The table below shows the valid operators that our application currently support.
```

Each query will consume the previous token and returns the next available token. +
This is similar to how `java.util.Scanner` works.

===== Design Consideration

After many rounds of experiment with `StringTokenizer` provided by native java and `ArgumentMultimap`, we found 3 main

> issues which they do not provide.

Reviewer 3: Capitalise `Java`.
Also, you need to rephrase the second half of the sentences.

Reviewer 3: Here you mentioned 3 issues, but there are only two below!

. `ArgumentMultimap` do not keep track of the order of each delimiter. They will only track if the delimiter provided

> exists.

Reviewer 3: `do` -> `does`

. `StringTokenizer` default delimiter is whitespace, although we could change the delimiter and parse multiple
delimiters. It does not suit our situation. +
e.g. `find name/AAAA code/BBBB`. +
This will return us 1 token. `name/AAAA code/BBBB`, which we could not use to check due to the missing `operator`. +
We need the tokenizer to return us 2 tokens `name/AAAA` and `code/BBBB` in order for us to know that the expression

> was invalid due to the missing `operator`

Reviewer 3: Having worked on this with you on the `StringTokenizer`, I think this could be better phrased.
Discuss more about how we still need to do additional parsing after using `StringTokenizer` to facilitate the execution of Shunting Yard.

> As such, our `Tokenizer` aim to solve these 2 issues.

Reviewer 3: `aim` -> `aims`.
Perhaps mentioned that we developed a custom `Tokenizer` that aims to solve these two issues.

> The table below shows the different of our `Tokenizer` vs `StringTokenizer` and `ArgumentMultimap`

Reviewer 3: `different` -> `differences`.

|====
| *`Tokenizer`* | *`StringTokenizer`* | *`ArgumentMultimap`*
| Respect the order of delimiters. | Respect the order of delimiters | Only keep track which delimiters is present.
| Support multiple delimiters. | Support only multiple delimiters. | Support multiple delimiters.
| More flexible in splitting of token | Only split token based on delimiters | Does not split into token.
|====

==== Operator

This is represented by the class `pwe.planner.logic.parser.Operator` and define all valid operators to be used in
`BooleanExpressionParser`.

The table below shows the valid operators our application currently support.

|====
| *Operator* | *Description* | *Precedence*
| `&&` | Logical "AND" operation (both conditions A AND B must match) | Highest
| `\|\|` | Logical "OR" operation (either conditions A OR B must match)| Lowest

> |====

Reviewer 3: Since parenthesis is also considered as an operator, perhaps include it and talk about it briefly here.

.Parser interactions
image::parserSequenceDiagram.png[width="650"]

> When the `FindCommandParser` received the provided user argument, after doing it checks, it will pass the argument to

Reviewer 2: Maybe you may want to elaborate more on what _checks_ is carried out?
```suggestion
When `FindCommandParser` receives the provided user argument, it will carry out checks and pass the argument to
```

[TIP]
====
See <<Tokenizer>> for more details regarding the tokenizer.
====

The code below shows how the predicate for each prefix is created.

```java
ArgumentMultimap argMultimap =
        ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_CODE, PREFIX_CREDITS);
KeywordsPredicate predicate = null;
if (argMultimap.getValue(PREFIX_NAME).isPresent()) {
    String nameKeyword = ParserUtil.parseName(argMultimap.getValue(PREFIX_NAME).get()).toString();
    predicate = new NameContainsKeywordsPredicate(List.of(nameKeyword));
} else if (argMultimap.getValue(PREFIX_CODE).isPresent()) {
    String codeKeyword = ParserUtil.parseCode(argMultimap.getValue(PREFIX_CODE).get()).toString();
    predicate = new CodeContainsKeywordsPredicate(List.of(codeKeyword));
} else if (argMultimap.getValue(PREFIX_CREDITS).isPresent()) {
    String creditKeyword = ParserUtil.parseCredits(argMultimap.getValue(PREFIX_CREDITS).get()).toString();
    predicate = new CreditsContainsKeywordsPredicate(List.of(creditKeyword));
} else {
    throw new ParseException(
            String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
}
return predicate;
```

> ==== Design Consideration

Reviewer 2: ```suggestion
==== Design Considerations
```

> This sections share details the design considerations we made when enhancing the existing `find` feature.

Reviewer 2: ```suggestion
This section shares the design considerations we went through during the enhancement of the existing `find` feature.
```

===== Aspect: Parsing of composite predicate

> The table below shows the comparison of different approach we could take.

Reviewer 2: ```suggestion
The table below shows comparisons between the two implementations that could have been made.
```

[cols="30%,35%,35%"]
|=====
|*Approach* | *Pros* | *Cons*|

> *1.* Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard] that parse the complex

Reviewer 2: ```suggestion
*1.* Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard] that parses the complex
```

boolean expression and returns a composite predicate. |

> Find command can be very flexible and scale with the application data.

Reviewer 2: Sounds quite vague.
```suggestion
Find command can be very flexible. It can work with many arguments to search for the specific modules that the user wants.
```

a|

> * Requires a lot of tests to ensure expected behaviour

Reviewer 2: ```suggestion
* Requires many tests to ensure expected behaviours
```

> * Extra work required to integrate a new logic parser into `FindCommandParser` and may cause build problems if not

Reviewer 2: ```suggestion
* Extra work is required to integrate a new logic parser into `FindCommandParser`. It may result in build problems if it is not
```

implemented properly. |
*2.* Do an implicit logical `OR` for every predicate. |
Very easy to implement a|
* Returns a very huge set of results when the number of keywords increase.
* Find command will be restricted to logical `OR` results.
|=====

> After weighing both approach pros and cons, we decided to go with the *approach 1*. +

Reviewer 2: ```suggestion
After weighing both pros and cons, we decided to go with the *approach 1*. +
```

> This is because we are expecting many common names between a module available in the university curriculum. If the

Reviewer 2: ```suggestion
This is because we are expecting many similar names between modules available in the university curriculum. If the
```

> user could only search with an implicit logical `OR`, they would not be able to find a module which suit them

Reviewer 2: ```suggestion
user could only search with an implicit logical `OR`, the user would not be able to find a module which suits the user's needs
```

> effectively. Which drastically reduce the effectiveness of the `find` command.

Reviewer 2: ```suggestion
effectively. This can drastically reduce the effectiveness of the `find` command.
```

.Parser interactions
image::parserSequenceDiagram.png[width="650"]

When the `FindCommandParser` received the provided user argument, after doing it checks, it will pass the argument to

>  `BooleanExpressionParser` which will initialise a new tokenizer which extract all argument as token. +

Reviewer 3: ```
initialise a new tokenizer which extract all argument as token
```
should be:
```
initialize a new `Tokenizer` that extracts the arguments as tokens
```

> With each token, the `BooleanExpressionParser` will then create a predicate for each prefix.

Reviewer 3: This should be:
```
`BooleanExpressionParser` will create a `Predicate` based on the `Prefix` in each token. If the token is an `Operator`, `BooleanExpressionParser` will apply the operator on two `Predicate` to combine them into a composite `Predicate.

[TIP]
====
See <<Tokenizer>> for more details regarding the tokenizer.
====

The code below shows how the predicate for each prefix is created.

```java
ArgumentMultimap argMultimap =
        ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_CODE, PREFIX_CREDITS);
KeywordsPredicate predicate = null;
if (argMultimap.getValue(PREFIX_NAME).isPresent()) {
    String nameKeyword = ParserUtil.parseName(argMultimap.getValue(PREFIX_NAME).get()).toString();
    predicate = new NameContainsKeywordsPredicate(List.of(nameKeyword));
} else if (argMultimap.getValue(PREFIX_CODE).isPresent()) {
    String codeKeyword = ParserUtil.parseCode(argMultimap.getValue(PREFIX_CODE).get()).toString();
    predicate = new CodeContainsKeywordsPredicate(List.of(codeKeyword));
} else if (argMultimap.getValue(PREFIX_CREDITS).isPresent()) {
    String creditKeyword = ParserUtil.parseCredits(argMultimap.getValue(PREFIX_CREDITS).get()).toString();
    predicate = new CreditsContainsKeywordsPredicate(List.of(creditKeyword));
} else {
    throw new ParseException(
            String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
}
return predicate;
```

==== Design Consideration

> This sections share some design consideration we encounter when enhancing the existing `find` feature.

Reviewer 3: `share some design consideration we encounter` -> `details the design considerations we made`

===== Aspect: Parsing of composite predicate

[NOTE]
Currently we are using choice 2 because this provide our users with a more flexible way of finding modules in our

> application.

Reviewer 3: Might want to reorder the two points and remove this note.

- Approach 1: Do an implicit logical `OR` for every predicate.

[cols="20%,<80%"]
|======
|*Pros*| Very easy to implementation
|*Cons* a|

> * Returns a very huge set of results when the keywords increase.

Reviewer 3: `keywords increase` -> `number of keywords increase`.

> * Find command will be restricted to logical `OR` results.

Reviewer 3: Could be better phrased.

|======

- Approach 2 (current choice): Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting
Yard]

>  that can parse complex boolean expression and composite the predicate.

Reviewer 3: You can't just `composite the predicate`. `Composite` isn't an action verb!

[cols="20%,<80%"]
|=====

> | *Pros* | Find command can be very flexible as per user wish.

Reviewer 3: Could go into a bit more details here.

> | *Cons* a|

Reviewer 3: Remove `a`.

Author: This is intended.
 
```
a| 
* point 1
* point 2
```

Allows the cells to have bullets point.

* Requires a lot of tests to ensure expected behaviour
* Extra work required to integrate a new logic parser into `FindCommandParser` and may cause build problems if not
implemented properly.
|=====

> After weighing both approach pros and cons, we have decided to go with the *approach 2*. +

Reviewer 3: `we have decided` -> `we decided`.

This is because we are expecting many common names between a module available in the university curriculum. If the
user could only search with an implicit logical `OR`, they would not be able to find a module which suit them

> effectively. Which defeat the purpose of our application.

Reviewer 3: `Which defeat the purpose of our application` is a strong statement.
Perhaps you could say: 
```
drastically reduce the effectiveness of the `find` command
```

[TIP]
====
See <<Tokenizer>> for more details regarding the tokenizer.
====

The code below shows how the predicate for each prefix is created.

```java
ArgumentMultimap argMultimap =
        ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_CODE, PREFIX_CREDITS);
KeywordsPredicate predicate = null;
if (argMultimap.getValue(PREFIX_NAME).isPresent()) {
    String nameKeyword = ParserUtil.parseName(argMultimap.getValue(PREFIX_NAME).get()).toString();
    predicate = new NameContainsKeywordsPredicate(List.of(nameKeyword));
} else if (argMultimap.getValue(PREFIX_CODE).isPresent()) {
    String codeKeyword = ParserUtil.parseCode(argMultimap.getValue(PREFIX_CODE).get()).toString();
    predicate = new CodeContainsKeywordsPredicate(List.of(codeKeyword));
} else if (argMultimap.getValue(PREFIX_CREDITS).isPresent()) {
    String creditKeyword = ParserUtil.parseCredits(argMultimap.getValue(PREFIX_CREDITS).get()).toString();
    predicate = new CreditsContainsKeywordsPredicate(List.of(creditKeyword));
} else {
    throw new ParseException(
            String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
}
return predicate;
```

==== Design Considerations

> This sections shares the design considerations we went through during the enhancing the existing `find` feature.

Reviewer 2: ```suggestion
This section shares the design considerations we went through during the enhancement of the existing `find` feature.
```

===== Aspect: Parsing of composite predicate
The table below shows comparisons between the two approaches that could have been made.
[cols="30%,35%,35%"]
|=====
|*Approach* | *Pros* | *Cons*|
*1.* Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard] that parses the complex
boolean expression and returns a composite predicate. |
Find command can be very flexible. It can work with multiple parameters to search for the specific modules that the
user wants.
a|
* Requires many tests to ensure expected behaviours
* Extra work is required to integrate a new logic parser into `FindCommandParser`. It may result in  build problems if
 it is not implemented properly.
 |*2.* Do an implicit logical `OR` for every predicate.
 |Very easy to implement a|
* Returns a very huge set of results when the number of keywords increase.
* Find command will be restricted to logical `OR` results.
|=====

> After weighing both pros and cons, we decided to go with the *approach 1*. +

Reviewer 2: ```suggestion
After weighing both pros and cons, we decided to go with *approach 1*. +
```

This is because we are expecting many similar names between modules in the university curriculum. If the

> user could only search with an implicit logical `OR`, the user would not be able to find a module which suit the

Reviewer 2: ```suggestion
user could only search with an implicit logical `OR`, the user would not be able to find modules which suit the
```
